Add error for virtual member function without self

dwblaikie opened the pull request:
This tripped over a lowering crash when a member function with self was
declared-but-not-defined, so that's why some test cases were updated to
have (empty) function definitions.

I'll follow-up with/look into a fix for the
self-declared-but-not-defined cases separately.

dwblaikie committed:
Add error for virtual member function without self

This tripped over a lowering crash when a member function with self was
declared-but-not-defined, so that's why some test cases were updated to
have (empty) function definitions.

I'll follow-up with/look into a fix for the
self-declared-but-not-defined cases separately.

jonmeow commented on toolchain/check/function.h:
```diff
@@ -33,7 +33,7 @@ auto CheckFunctionTypeMatches(
     const SemIR::Function& prev_function,
     SemIR::SpecificId prev_specific_id = SemIR::SpecificId::None,
-    bool check_syntax = true) -> bool;
+    bool check_syntax = true, bool allow_self_type_mismatch = false) -> bool;
```
This is feeling like a lot of optional parameters, but looking at current call sites of both this and CheckRedeclParamsMatch, it seems most pass none of the optional params. WDYT of changing both to have one overload with none of the optional params, and one overload with all params required? I think that'd make you change impl.cpp+handle_impl.cpp to specify allow_self_type_mismatch=false.

Note on CheckRedeclParamsMatch that might also nudge towards shifting the order of diagnose so that parameter clustering matched CheckFunctionTypeMatches (whereas right now `diagnose` is in the middle).

dwblaikie replied:
ah, sorry, missed responding to/addressing this comment. Will do so shortly.

dwblaikie replied:
Added overloads and reordered the parameters in 69bac1b78255829d79dfa5e20a4b03c8b15f841d

jonmeow commented on toolchain/check/function.h:
```diff
@@ -33,7 +33,7 @@ auto CheckFunctionTypeMatches(
     const SemIR::Function& prev_function,
     SemIR::SpecificId prev_specific_id = SemIR::SpecificId::None,
-    bool check_syntax = true) -> bool;
+    bool check_syntax = true, bool allow_self_type_mismatch = false) -> bool;
```
check_syntax and allow_self_type_mismatch seem like they're addressing a similar issue, but with different naming schemes. Had you considered either renaming check_syntax to allow_syntax_mismatch, or renaming allow_self_type_mismatch to check_self_type to match?

Separately, might also be worth a comment here and on CheckRedeclParamsMatch about at least the new param, since we're building up a number.

dwblaikie replied:
Fair point on the rename (went with `check_self` because it's more than just the type that is/isn't being checked - like whether it's written with `addr` or not, for instance) in 9a5bb5574b50dcef64e6a940f28a3d88d9a8046e

Added comments for `check_syntax` and `check_self` in d88c8b4e2a1e23bf7cde2ad436fd2a6aeb3baceb

jonmeow commented on toolchain/check/merge.cpp:
```diff
@@ -233,7 +233,14 @@ static auto CheckRedeclParam(Context& context, bool is_implicit_param,
   auto prev_param_type_id = SemIR::GetTypeInSpecific(
       context.sem_ir(), prev_specific_id, prev_param_pattern.type_id());
-  if (!context.types().AreEqualAcrossDeclarations(new_param_pattern.type_id(),
+  if ((!allow_self_type_mismatch ||
```
This is a pretty complex `if`, maybe add a brief implementation comment?

e.g.:

```suggestion
  // Make sure the param type matches, unless it's a `self` param and a different type is allowed.
  if ((!allow_self_type_mismatch ||
```

dwblaikie replied:
Now that the name check is pulled out separately, maybe not worth the comment? (at a similar level of complexity as the other checks in this function - and easier to read when it's an affirmative check "allow self type mismatch && new name == self && prev name == self, then return true/skip checking the type")

jonmeow commented on toolchain/check/merge.cpp:
```diff
@@ -459,21 +467,23 @@ static auto CheckRedeclParamSyntax(Context& context,
+                         allow_self_type_mismatch)) {
     return false;
   }
   if (!CheckRedeclParams(context, new_entity.loc, new_entity.param_patterns_id,
```
```suggestion
  // Don't forward `allow_self_type_mismatch` here because it's extra cost, and `self` is only allowed in implicit params.
  if (!CheckRedeclParams(context, new_entity.loc, new_entity.param_patterns_id,
```

jonmeow commented on toolchain/check/testdata/class/virtual_modifiers.carbon:
```diff
@@ -13,13 +13,13 @@
 
 base class Base {
-  virtual fn H();
+  virtual fn H[self: Self]();
```
All these tests are adding `[self: Self]`. Should we have some tests of things like `[addr self: Self*]` and `[addr self: Base*]` (so that it's not only testing `Self`)?

dwblaikie replied:
Ah, indeed - added 3ba49885c51186ac1fcf32f1ef545dbc0c833514

The spec seems a bit vague/doesn't mention whether it's OK to mismatch `self` and `addr self` between an original (`virtual` or `abstract`) function, and an `impl`. But to me, at least, it seems reasonable to allow mismatch.

Though I did discover in the language design docs, discussion about covariant return types - though that's also not fully spec'd, just mentioned as an aside in another section. (covariant parameters could also be supported/discussed) - but I'll leave those for follow-up changes. I note it here because it means the redecl matching logic will change even moreso for these virtual function redeclarations.

dwblaikie replied:
Oh, realized the covariant cases are already handled by the `check_syntax = false` parameter, which handles implicit conversions from parameters/to the result.

Actually, I'm wondering if `check_self` can be generalized and heaped under the `check_syntax` bucket (looking at how this behavior might differ/be similar with interface impls)

jonmeow replied:
I think `self` vs `addr self` should [require a] match since that's essentially changing between `const` and non-`const` in C++ terms. (meaningfully different signatures, even for inheritance -- and changing from `self` to `addr self` is a really significant contract change)

jonmeow replied:
Yeah, wasn't totally sure about `check_syntax`, but felt like impls might be slightly differently constrained.

dwblaikie replied:
Ah, right - checking `addr` done in c82572a10cf118b5a6d4385965af5b619055e8c8

From a conversation with @zygoloid, @chandlerc - it seems like this checking code might be replaced with creating a thunk in SemIR that would synthesize the desired call, and error from there. For self, various changes could be valid (eg: a virtual function that takes `addr self` could be impl'd with a function that takes `self` (same way normal a member function taking `addr self` could call a member function taking `self`, but not the inverse))

But hopefully this level of self checking will suffice for now - I might try the thunk thing next, though it seems a bit like a big lift since synthesizing that amount of SemIR apparently doesn't have precedent so far. (not sure what the diagnostic experience will be like, etc)

jonmeow commented on toolchain/check/handle_function.cpp:
```diff
@@ -278,6 +278,11 @@ static auto BuildFunctionDecl(Context& context,
 
+  if (virtual_modifier != SemIR::Function::VirtualModifier::None &&
+      !self_param_id.has_value()) {
+    CARBON_DIAGNOSTIC(VirtualWithoutSelf, Error, "virtual non-instance method");
```
```suggestion
    CARBON_DIAGNOSTIC(VirtualWithoutSelf, Error, "virtual class function");
```

Terminology comes from:
https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/classes.md?plain=1#L850

> Member functions can either be class functions or methods. Class functions are
> members of the type, while methods can only be called on instances.

(on my mind because we were recently discussing this terminology)

dwblaikie committed:
Rephrase "non-instance method" to "class function"

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

dwblaikie committed:
Add comment describing why `allow_self_type_mismatch` doesn't apply to explicit params

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

dwblaikie committed:
Line wrap comment

Co-authored-by: Carbon Infra Bot <carbon-external-infra@google.com>

dwblaikie committed:
Update tests

jonmeow commented on toolchain/check/merge.cpp:
```diff
@@ -233,7 +233,14 @@ static auto CheckRedeclParam(Context& context, bool is_implicit_param,
+       (SemIR::Function::GetNameFromPatternId(context.sem_ir(),
+                                              new_param_pattern_id) !=
+            SemIR::NameId::SelfValue ||
+        SemIR::Function::GetNameFromPatternId(context.sem_ir(),
```
Sorry, this is bothering me as I think about how this works. Lines 280-284 are fundamentally about querying the parameter name. Is there a way that the result of the name fetch could be reused for this logic?

dwblaikie replied:
ae526d1591fb0003d781267bd7914dc2a46e501f - there's still a bunch of work that's done inside `GetNameFromPatternId` that's somewhat redundant with parts later in the function (the code that looks through `AddrPattern` and `ValueParamPattern`, I think) - but I struggled to find a nice way to handle both cases - where the parameters colud be quite different (addr, non-addr, etc) so long as they are both named `self`, with the "need to check they're the same as we look through the layers of abstraction".

Happy to riff on this if you reckon it's worth it/have any suggestions/etc.

jonmeow replied:
Note, not sure if it's easier to merge with my above comment that the shape of `self` probably should match... i.e., maybe we can just move the type comparison to be after the `addr` unwrap and name comparison?

dwblaikie replied:
Yeah, rolled that into the other change ( c82572a10cf118b5a6d4385965af5b619055e8c8 ).

dwblaikie committed:
Support `addr self`, including mismatches of `addr self` and `self`

dwblaikie committed:
Share name lookup results

dwblaikie committed:
Rename (and invert) `allow_self_type_mismatch` to `check_self`

dwblaikie committed:
Document the `check_*` parameters

dwblaikie committed:
Refactor CheckRedeclParamsMatch and CheckFunctionTypeMatches

Make them both have an overload that takes all the (previously) optional
parameters mandatorily, and one that takes none of them.

And reorder the {check_syntax, diagnose, check_self} to put diagnose
first so the check_* go together.

dwblaikie committed:
Check `addr` matches between redeclarations.

jonmeow commented on toolchain/check/merge.h:
```diff
@@ -96,11 +96,20 @@ struct DeclParams {
+                                   const DeclParams& new_entity,
+                                   const DeclParams& prev_entity) -> bool {
+  return CheckRedeclParamsMatch(context, new_entity, prev_entity,
+                                SemIR::SpecificId::None, /*diagnose=*/true,
+                                /*check_syntax=*/true,
+                                /*check_self=*/true);
```
```suggestion
                                SemIR::SpecificId::None, /*check_syntax=*/true,
                                /*check_self=*/true, /*diagnose=*/true);
```

Note because of the `/*` clang-format won't wrap parameters when they could be. But this is an ordering comment, which clang-tidy might also catch. :)

jonmeow commented on toolchain/check/merge.cpp:
```diff
@@ -231,6 +231,27 @@ static auto CheckRedeclParam(Context& context, bool is_implicit_param,
+    }
+  }
+
+  auto new_name_id = SemIR::Function::GetNameFromPatternId(
+      context.sem_ir(), new_param_pattern_id);
+  auto prev_name_id = SemIR::Function::GetNameFromPatternId(
+      context.sem_ir(), prev_param_pattern_id);
```
Small at this point, but could this go back to the prior:

```suggestion
  auto new_name_id = context.entity_names().Get(
      new_param_pattern.As<SemIR::AnyBindingPattern>().entity_name_id).name_id;
  auto prev_name_id = context.entity_names().Get(
      prev_param_pattern.As<SemIR::AnyBindingPattern>().entity_name_id).name_id;
```

i.e., because `GetNameFromPatternId` includes more logic, and if we don't need it, then this is efficient.

jonmeow reviewed (approved):
Thanks! Note some failing tests, plus a couple minor comments, but I think this is basically good to merge.

dwblaikie committed:
Trivial test update

dwblaikie committed:
Fix parameter name to match implementation/callers and line wrapping

dwblaikie committed:
Avoid `GetNameFromPatternId` with all its extra generality

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

dwblaikie committed:
Move the ValueParamPattern checking to go before name lookup
